File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/ParentWalker.java
package com.greentree.commons.graph.algorithm.walk;

import com.greentree.commons.graph.RootTree;

public class ParentWalker<V> implements RootTreeWalker<V> {
	
	private final RootTree<? extends V> tree;
	
	public ParentWalker(RootTree<? extends V> tree) {
		this.tree = tree;
	}
	
	@Override
	public void visit(V start, GraphVisitor<? super V> visitor) {
		walk(visitor, start);
	}
	
	private void walk(GraphVisitor<? super V> visitor, V v) {
		visitor.startVisit(v);
		final var to = tree.getParent(v);
		walk(visitor, to, v);
		visitor.endVisit(v);
	}
	
	private void walk(GraphVisitor<? super V> visitor, V v, V parent) {
		visitor.startVisit(parent, v);
		final var to = tree.getParent(v);
		walk(visitor, to, v);
		visitor.endVisit(parent, v);
	}
	
	@Override
	public RootTree<? extends V> graph() {
		return tree;
	}
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/ParentWalker.kt
package com.greentree.commons.graph.algorithm.walk

import com.greentree.commons.graph.RootTree
import com.greentree.commons.graph.algorithm.walk.vertex.VertexVisitor

class ParentWalker<V : Any>(private val tree: RootTree<V>) :
	RootTreeWalker<V> {

	override fun visit(start: V, visitor: VertexVisitor<V>) {
		walk(visitor, start)
	}

	override fun graph(): RootTree<V> {
		return tree
	}

	override fun visit(visitor: VertexVisitor<V>) {
		super.visit(visitor)
	}

	private fun walk(visitor: VertexVisitor<V>, v: V) {
		visitor.startVisit(v)
		val to = tree.getParent(v)
		walk(visitor, to, v)
		visitor.endVisit(v)
	}

	private fun walk(visitor: VertexVisitor<V>, v: V, parent: V) {
		visitor.startVisit(parent, v)
		val to = tree.getParent(v)
		walk(visitor, to, v)
		visitor.endVisit(parent, v)
	}
}
